NitroxClient/Communication/ClientBridge.cs

```diff
@@ -0,0 +1,249 @@
+            }
+            catch (Exception ex)
+            {
+                TryStopClient();
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+        }
+
+        public void ClaimReservation()
+        {
+            try
+            {
+                if (CurrentState != ClientBridgeState.Reserved)
+                {
+                    throw new ProhibitedClaimAttemptException();
+                }
+
+                var packet = new ClaimPlayerSlotReservation(correlationId, ReservationKey);
+                serverClient.Send(packet);
+
+                CurrentState = ClientBridgeState.Connected;
+            }
+            catch (Exception ex)
+            {
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+        }
+
+        public void HandleRejectedReservation(string correlationId, ReservationRejectionReason rejectionReason)
+        {
+            if (CurrentState == ClientBridgeState.Disconnected)
+            {
+                return;
+            }
+
+            try
+            {
+                if (CurrentState != ClientBridgeState.WaitingForRerservation)
+                {
+                    throw new ProhibitedReservationRejectionException();
+                }
+
+                ValidateCorrelationId(correlationId);
+                ConfirmCorrelationId(correlationId);
+                ValidateRejectionReason(rejectionReason);
+
+                ReservationKey = null;
+                ReservationRejectionReason = rejectionReason;
+
+                CurrentState = ClientBridgeState.ReservationRejected;
+            }
+            catch (Exception ex)
+            {
+                TryStopClient();
+                CurrentState = ClientBridgeState.Failed;
+                throw ex;
+            }
+        }
+
+        public void Send(Packet packet)
+        {
+            if (CurrentState != ClientBridgeState.Connected)
+            {
+                throw new ProhibitedSendAttemptException();
+            }
+
+            serverClient.Send(packet);
+        }
+
+        public PacketSuppression<T> Suppress<T>()
+        {
+            return new PacketSuppression<T>(suppressedPacketsTypes);
+        }
+
+        private void TryStopClient()
+        {
+            if (serverClient.IsConnected)
+            {
+                serverClient.Stop();
+            }
+        }
+
+        private void RequestReservation(string playerName)
+        {
+            if (CurrentState == ClientBridgeState.WaitingForRerservation ||
+                CurrentState == ClientBridgeState.Reserved ||
+                CurrentState == ClientBridgeState.Connected)
+            {
+                throw new ProhibitedConnectAttemptException();
+            }
+
+            var reservePlayerSlot = new ReservePlayerSlot(correlationId, playerName);
+            serverClient.Send(reservePlayerSlot);
+
+            PlayerId = playerName;
```

Comment:
We should decouple playerId from playerName.  Ideally playerId was a guid and player name will change server side if there is a duplicate... i.e.: Sunrunner -> Sunrunner#2

Reply:
I'd like to wait to do this until we find a better place to store that information. I don't feel like this is the appropriate place at all and it's just kind of here for legacy support at the moment.

Reply:
And if so, do we want to re-use the `ReservationKey` for `PlayerId`? Would that provide any complications? IIRC you mentioned somewhere else that this key coul be used for persistent members: that way the position, inventory and other related information can be stored on the server by player. Other related info being extra privileges (related to join-reservations), for instance.

Reply:
The reservation and correlation keys needs to be secret and generated on the fly or else it could be considered a security vulnerability. It is kind of like not salting your hash =)